### PR TITLE
Dir host wrapper

### PR DESCRIPTION
**Pull request by tbosch**
The first 7 commits are from https://github.com/angular/angular/pull/12445

### CODE REVIEW

File: modules/@angular/compiler/src/compiler_util/identifier_util.ts [hunk 1]
```diff
@@ -6,9 +6,11 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {SecurityContext} from '@angular/core';
```
Comment (vicb) [inline modules/@angular/compiler/src/compiler_util/identifier_util.ts]: remove

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 2]
```diff
@@ -184,3 +206,74 @@ function addCheckInputMethod(input: string, builder: DirectiveWrapperBuilder) {
...
+      public errors: ParseError[]) {}
+}
+
+function parseHostBindings(
+    dirMeta: CompileDirectiveMetadata, exprParser: Parser,
+    schemaRegistry: ElementSchemaRegistry): ParseResult {
+  const errors: ParseError[] = [];
+  const parser =
+      new BindingParser(exprParser, DEFAULT_INTERPOLATION_CONFIG, schemaRegistry, [], errors);
+  var sourceFileName = dirMeta.type.moduleUrl ?
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: const

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 3]
```diff
@@ -184,3 +206,74 @@ function addCheckInputMethod(input: string, builder: DirectiveWrapperBuilder) {
...
+  const warnings = parseErrors.filter(error => error.level === ParseErrorLevel.WARNING);
+  const errors = parseErrors.filter(error => error.level === ParseErrorLevel.FATAL);
+
+  if (warnings.length > 0) {
+    this._console.warn(`Template parse warnings:\n${warnings.join('\n')}`);
+  }
+
+  if (errors.length > 0) {
+    const errorString = errors.join('\n');
+    throw new Error(`Template parse errors:\n${errorString}`);
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: directive host...

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 4]
```diff
@@ -240,42 +240,33 @@ export class BindingParser {
...
-        this._reportError(errorMsg, boundProp.sourceSpan);
-      }
     } else {
       if (parts[0] == ATTRIBUTE_PREFIX) {
         boundPropertyName = parts[1];
         this._validatePropertyOrAttributeName(boundPropertyName, boundProp.sourceSpan, true);
-        // NB: For security purposes, use the mapped property name, not the attribute name.
-        const mapPropName = this._schemaRegistry.getMappedPropName(boundPropertyName);
-        securityContext = this._schemaRegistry.securityContext(elementName, mapPropName);
+        securityContexts = possibleSecurityContexts(
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: verb

File: modules/@angular/compiler/src/compiler_util/identifier_util.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/compiler_util/identifier_util.ts]: Done

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Done

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Done

File: modules/@angular/compiler/src/schema/dom_element_schema_registry.ts [hunk 5]
```diff
@@ -8,6 +8,8 @@
 
 import {CUSTOM_ELEMENTS_SCHEMA, Injectable, NO_ERRORS_SCHEMA, SchemaMetadata, SecurityContext} from '@angular/core';
 
+import {CssSelector} from '../selector';
```
Comment (tbosch) [inline modules/@angular/compiler/src/schema/dom_element_schema_registry.ts]: Done

File: modules/@angular/compiler/src/template_parser/binding_parser.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: Done

File: modules/@angular/compiler/src/template_parser/binding_parser.ts [hunk 6]
```diff
@@ -429,3 +419,24 @@ export class PipeCollector extends RecursiveAstVisitor {
...
+export function calcPossibleSecurityContexts(
+    registry: ElementSchemaRegistry, selector: string, propName: string,
+    isAttribute: boolean): SecurityContext[] {
+  const ctxs: SecurityContext[] = [];
+  CssSelector.parse(selector).forEach((selector) => {
+    const elementNames = selector.element ? [selector.element] : registry.allKnownElementNames();
+    const notElementNames =
+        new Set(selector.notSelectors
+                    .filter(
+                        selector => selector.element && selector.attrs.length === 0 &&
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: selector.isElementSelector()

Comment (tbosch) [inline modules/@angular/compiler/src/template_parser/binding_parser.ts]: Thanks!

**Comment by vsavkin** [comment]
Merged https://github.com/angular/angular/commit/178fb79b5cafe36c993bb1f9f9a63e8bac162c66

**Pull request closed without merging**